pf/proto/protov6.go

```diff
@@ -0,0 +1,151 @@
+		ctx:    ctx,
+		server: server,
+		getSchema: sync.OnceValues(func() (*tfprotov6.GetProviderSchemaResponse, error) {
+			return server.GetProviderSchema(ctx, &tfprotov6.GetProviderSchemaRequest{})
+		}),
+	}
+}
+
+// Replace the inner provider with a new one.
+func (p *Provider) Replace(newServer tfprotov6.ProviderServer) {
+	n := New(p.ctx, newServer)
+	newProvider := n.(*Provider)
+	p.getSchema = newProvider.getSchema
+	p.server = newProvider.server
+}
+
+// Provider provides a shim from [tfprotov6.ProviderServer] to [shim.Provider].
+//
+// To create a Provider, use [New]. A zero value Provider will panic on most operations.
+//
+// As much as possible, Provider is designed to be a wrapper type. It is cheap to create
+// and use because it doesn't call any methods on the underlying server until it needs to.
+//
+// Provider is part of an internal API and does not have any compatibility guarantees. It
+// may be changed or removed in future updates.
+type Provider struct {
+	// The underlying server.
+	server tfprotov6.ProviderServer
+
+	// A cached GetSchema on the underlying server.
+	getSchema func() (*tfprotov6.GetProviderSchemaResponse, error)
+
+	// A cached copy of the ctx that the provider was originally created with.
+	//
+	// Used during Replace.
+	ctx context.Context
+}
+
+// Get access to the underlying sever used in Provide
+func (p Provider) Server(context.Context) (tfprotov6.ProviderServer, error) {
+	return cachedSchemaProvider{p.server, p.getSchema}, nil
+}
+
+func (p Provider) Config(context.Context) (tftypes.Object, error) {
+	v, err := p.getSchema()
+	if err != nil {
+		return tftypes.Object{}, err
+	}
+	return v.Provider.ValueType().(tftypes.Object), nil
+}
+
+type cachedSchemaProvider struct {
+	tfprotov6.ProviderServer
+
+	getSchema func() (*tfprotov6.GetProviderSchemaResponse, error)
+}
+
+func (p cachedSchemaProvider) GetProviderSchema(
+	context.Context, *tfprotov6.GetProviderSchemaRequest,
+) (*tfprotov6.GetProviderSchemaResponse, error) {
+	return p.getSchema()
+}
+
+func (p Provider) Schema() shim.SchemaMap {
+	v, err := p.getSchema()
+	if err != nil {
+		tfbridge.GetLogger(p.ctx).Error(err.Error())
+		return nil
+	}
+	if provider := v.Provider; provider != nil {
+		return blockMap{provider.Block}
+	}
+	return blockMap{&tfprotov6.SchemaBlock{}}
+}
+
+func (p Provider) ResourcesMap() shim.ResourceMap {
+	v, err := p.getSchema()
+	if err != nil {
+		tfbridge.GetLogger(p.ctx).Error(err.Error())
+		return nil
+	}
+	return resourceMap(v.ResourceSchemas)
+}
+
+func (p Provider) DataSourcesMap() shim.ResourceMap {
+	v, err := p.getSchema()
+	if err != nil {
+		tfbridge.GetLogger(p.ctx).Error(err.Error())
+		return nil
+	}
+	return resourceMap(v.DataSourceSchemas)
+}
+
+func filter[T any](m []T, keep func(T) bool) []T {
+	for i := len(m) - 1; i >= 0; i-- {
+		if keep(m[i]) {
+			continue
+		}
+
+		// If i is not the last element in the slice, put the last element in the
+		// slice into i.
+		if i < len(m)-1 {
```

Comment:
If this happens would we check the last element in the new position with `keep`?

I'd appreciate a test on this, not obviously correct.

Reply:
We wouldn't need to, since it would have already been checked. 

Doesn't matter though, the standard library has `slices.DeleteFunc`, which is what we should have used. I'm removing `filter` in place of `slices.DeleteFunc`.